src/main/java/xyz/nkomarn/harbor/listener/AfkListener.java

```diff
@@ -92,11 +92,22 @@ public void onLeave(PlayerQuitEvent event) {
      * Internal class for handling the task of checking player movement; Is a separate task so that we can cancel and restart it easily
      */
     private final class PlayerMovementChecker extends BukkitRunnable {
+        private int checksToMake = 0;
         @Override
         public void run() {
-            // This might cause lag if there are a TON of players on the server, but by keeping things minimalistic,
-            // we can try hard to reduce it. This also short-circuits if there are no players to check.
-            players.stream().filter(AfkPlayer::changed).forEach(afkPlayer -> afkProvider.updateActivity(afkPlayer.player));
+            if(players.isEmpty()){
+                checksToMake = 0;
+                return;
+            }
+
+            // We want every player to get a check every 20 ticks. Therefore we check 1/20th of the players
+            for (checksToMake += Math.ceil(players.size() / 20.0); checksToMake > 0 && !players.isEmpty(); checksToMake--) {
+                AfkPlayer afkPlayer = players.poll();
+                if (afkPlayer.changed()) {
+                    afkProvider.updateActivity(afkPlayer.player);
+                }
+                players.add(afkPlayer);
+            }
```

Comment:
just use a double :D This wont kill anyone and will make the logic simpler :D
Currently you will check a player in an unstable state. If one player is on the server the player will be checked every tick.
4 Players -> Every 5 Ticks
10 Players -> Every 10 Ticks
Only after 20 players online the splitting will have the effect I want.
That I used a double had the thought that I might want to check only a friction of the players and I will wait until I have a full player `1 > checksToMake`. this will allow empty runs when less then 20 players are online and when more then 20 players are online it will check in some runs more than one,
Doubles arent bad for loops :D

Otherwise we could just merge it and I will apply the optimizations by myself in a new PR. Your choice.

Or still the choice of the maintainers of this project.

```java
        if (players.isEmpty()) return;

        checksToMake += players.size() / 20.0;

        while (checksToMake > 1) {
            AfkPlayer afkPlayer = players.poll();
            if (afkPlayer.changed()) {
                playerManager.updateActivity(afkPlayer.player);
            }
            players.add(afkPlayer);
            checksToMake--;
        }
```

Reply:
Converted to double, using for loop syntax:
```java
 private final class PlayerMovementChecker extends BukkitRunnable {
        private double checksToMake = 0;
        @Override
        public void run() {
            if(players.isEmpty()){
                checksToMake = 0;
                return;
            }

            // We want every player to get a check every 20 ticks. Therefore we check 1/20th of the players
            for (checksToMake += players.size() / 20.0D; checksToMake > 0 && !players.isEmpty(); checksToMake--) {
                AfkPlayer afkPlayer = players.poll();
                if (afkPlayer.changed()) {
                    afkProvider.updateActivity(afkPlayer.player);
                }
                players.add(afkPlayer);
            }
        }
    }
```